### DISCUSSION TITLE
Trying to use my own dummy log receiver with OT collector using custom builder, keep getting this error

### DISCUSSION THREAD

**Discussion by AJ-54** [open-telemetry/opentelemetry-collector-contrib#3734] [Q&A]
I am not able to use my own receiver with OT. How can I resolve this error?
![image](https://user-images.githubusercontent.com/44471377/121124103-fabc3580-c841-11eb-8e4f-0483ff2568ec.png)
How can my receiver use the internal package and how should I import this receiver inside my `components.go` file of custom OT builder?

**Comment by SergeyKanzhelev** [comment]
Maybe check here: https://medium.com/opentelemetry/extending-the-opentelemetry-collector-with-your-own-components-64c10cf675db

CC: @jpkrohling

**Reply by AJ-54** [reply]
The problem is that internal package cannot be accessed outside the parent directory, so I placed my receiver inside the receiver folder where other contib receivers are present and to make custom build, I placed my custom build code folder inside parent as well but it says to me that usage of internal folder is not allowed.

**Reply by AJ-54** [reply]
I made some progress, I'll edit the question for better understanding of my problem

**Reply by bogdandrutu** [reply]
@djaglowski please help

**Reply by djaglowski** [reply]
> How can my receiver use the internal package

We can discuss extracting an API from the package and making this available outside the contrib collector. There are a small number of structs / methods that would need to be made available. It's ultimately up to the maintainers whether this would be acceptable, but here are a couple details we'd need to iron out:
- Where would be the appropriate place to move this? Possibly a new package called `logreceivercreator`?
- Probably, this API needs to be marked as experimental, and expectations should be that it will change.

**Reply by AJ-54** [reply]
@djaglowski For the purpose of developing the receiver locally, I renamed my modules to point at `github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dummyreceiver` and it is now able to access the internal API's locally. But this also means that the log receiver which is being created using the opentelemetry-log-collection library can only be used inside that GitHub repo.

**Reply by djaglowski** [reply]
The internal package was not explicitly meant for external consumption. However, as I've noted above, we should be able to extract a public interface from the internal package. This requires buy in from the maintainers, and will take some time to design & implement. I've created #3744 to discuss.

**Reply by AJ-54** [reply]
Thank you @djaglowski for raising this issue. Really appreciate it.

**Comment by jpkrohling** [comment]
Internal APIs aren't meant to be consumed by other components. What are you trying to use from that package?

**Reply by AJ-54** [reply]
HI @jpkrohling, Is there a better way to create my own log receiver? 
I see all the log receivers which are out there inside collector-contrib, namely  [filelogreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver), [tcplogreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/tcplogreceiver), [syslogreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/syslogreceiver) and [udplogreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/udplogreceiver) uses stanza and other operators from opentelemetry-log-collection library. If these are internal API's, won't that restrict contributions from other individuals?

**Reply by jpkrohling** [reply]
Sorry, I guess my general answer wasn't a good one for this case. There's apparently something specific to the log components that I'm not aware of.
